### ISSUE TITLE
OpenSlides behind Traefik2 Reverse Proxy. 

### ISSUE THREAD

**Issue by Redactlos3** [OpenSlides/OpenSlides#6109]
Hello all,
I am currently trying to put OpenSlides behind a Traefik proxy. Unfortunately, the two approaches I've followed both don't work. Has anyone had the same problem or has an idea to solve the problem? I don't care which of the two approaches works in the end. The main thing is that it works.

Thanks for helping.
Redactlos3

### Approach 1:
I first tried to route the port (9180:443) required by openslides_proxy via traefik. In doing so, I get an InternalServerError (HTTP 500).  

**docker-compose.yml:**
_The containers or settings not listed correspond to the configuration from docker-compose.yml.m4._

 ```
 proxy:
    container_name: os-proxy
    image: openslides/openslides-proxy:latest
    depends_on:
      - server
      - client
      - autoupdate
      - media
    environment:
        INSTANCE_DOMAIN: "10.0.0.10:9180"
        # EXTERNAL_HTTP_PORT: "9181"
        EXTERNAL_HTTPS_PORT: "9180"
        ALLOWED_HOSTS: ""
    labels:
      - "traefik.enable=true"
      ## HTTP Routers
      - "traefik.http.routers.openslides-rtr.entrypoints=https"
      - "traefik.http.routers.openslides-rtr.rule=Host(`openslides.example.de`)"
      ## Middlewares
      - "traefik.http.routers.openslides-rtr.middlewares=chain-no-auth@file" 
      ## HTTP Services
      - "traefik.http.routers.openslides-rtr.service=openslides-svc"
      - "traefik.http.services.openslides-svc.loadbalancer.server.scheme=https"
      - "traefik.http.services.openslides-svc.loadbalancer.server.port=8001"
    networks:
      - front
      - back
      - t2_proxy
    ports:
      - "9180:8001"
```

**traefik middleware configuration:**

```
[http.middlewares]

  [http.middlewares.chain-no-auth]
    [http.middlewares.chain-no-auth.chain]
      middlewares = [ "middlewares-rate-limit", "middlewares-secure-headers"]

  [http.middlewares.middlewares-rate-limit]
    [http.middlewares.middlewares-rate-limit.rateLimit]
      average = 100
      burst = 50

  [http.middlewares.middlewares-secure-headers]
    [http.middlewares.middlewares-secure-headers.headers]
      accessControlAllowMethods= ["GET", "OPTIONS", "PUT"]
      accessControlMaxAge = 100
      hostsProxyHeaders = ["X-Forwarded-Host"]
      sslRedirect = true
      stsSeconds = 63072000
      stsIncludeSubdomains = true
      stsPreload = true
      forceSTSHeader = true
#      frameDeny = true #overwritten by customFrameOptionsValue
      customFrameOptionsValue = "allow-from https:example.de" #CSP takes care of this but may be needed for organizr. 
      contentTypeNosniff = true 
      browserXssFilter = true 
#      sslForceHost = true # add sslHost and all of the 
#      sslHost = "example.de"
      referrerPolicy = "same-origin" 
#      Setting contentSecurityPolicy is more secure but it can break things. Proper auth will reduce the risk.
#      the below line also breaks some apps due to 'none' - sonarr, radarr, etc.
#      contentSecurityPolicy = "frame-ancestors '*.example.de:*';object-src 'none';script-src 'none';"
      featurePolicy = "camera 'none'; geolocation 'none'; microphone 'none'; payment 'none'; usb 'none';"
      [http.middlewares.middlewares-secure-headers.headers.customResponseHeaders]
        X-Robots-Tag = "none,noarchive,nosnippet,notranslate,noimageindex,"
        server = ""
```
### Approach 2:

I used the caddyfile as a guide and routed the paths to the corresponding ports. I also get an InternalServerError (HTTP 500).

**Traefik router and load balancer configuration.**

```
[http.routers]
  [http.routers.os-rtr-client]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`)"
      service = "os-svc-4200"
      middlewares = ["chain-no-auth"]

  [http.routers.os-rtr-media]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`) && Path(`/media/*`)"
      service = "os-svc-8000-m"
      middlewares = ["chain-no-auth"]

  [http.routers.os-rtr-srv01]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`) && Path(`/apps/*`)"
      service = "os-svc-8000"
      middlewares = ["chain-no-auth"]

  [http.routers.os-rtr-srv02]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`) && Path(`/reset/*`)"
      service = "os-svc-8000"
      middlewares = ["chain-no-auth"]

  [http.routers.os-rtr-srv03]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`) && Path(`/server-version.txt`)"
      service = "os-svc-8000"
      middlewares = ["chain-no-auth"]

  [http.routers.os-rtr-prometheus]
      entryPoints = ["https"]
      rule = "Host(`openslides.example.de`) && Path(`/prometheus/*`)"
      service = "os-svc-9090"
      middlewares = ["chain-no-auth"]

[http.services]
  [http.services.os-svc-4200]
    [http.services.os-svc-4200.loadBalancer]
      passHostHeader = true
      [[http.services.os-svc-4200.loadBalancer.servers]]
	scheme = "https"
        url = "https://10.0.0.10:9142" # internal docker port => 4200

  [http.services.os-svc-8002]
    [http.services.os-svc-8002.loadBalancer]
      passHostHeader = true
      [[http.services.os-svc-8002.loadBalancer.servers]]
        scheme = "https"
        url = "https://10.0.0.10:9182" # internal docker port => 8002

  [http.services.os-svc-8000]
    [http.services.os-svc-8000.loadBalancer]
      passHostHeader = true
      [[http.services.os-svc-8000.loadBalancer.servers]]
        scheme = "https"
        url = "https://10.0.0.10:9181" # internal docker port => 8000

  [http.services.os-svc-8000-m]
    [http.services.os-svc-8000-m.loadBalancer]
      passHostHeader = true
      [[http.services.os-svc-8000-m.loadBalancer.servers]]
        scheme = "https"
        url = "https://10.0.0.10:9180" # internal docker port => 8000

  [http.services.os-svc-9090]
    [http.services.os-svc-9090.loadBalancer]
      passHostHeader = true
      [[http.services.os-svc-9090.loadBalancer.servers]]
        scheme = "https"
        url = "https://10.0.0.10:9190" # internal docker port => 9090

```

**Shared ports docker-compose.yml:**
```
autoupdate:
    container_name: os-autoupdate
    image: openslides/openslides-autoupdate:latest
    ...
    ports:
       - "9082:8002"
    networks:
      - back

  media:
    container_name: os-media
    image: openslides/openslides-media:latest
    ...
    ports:
      - "9180:8000"

  server:
    container_name: os-server
    << : *default-osserver
    depends_on:
      - server-setup
    environment:
      << : *default-osserver-env
    ports:
      - "9181:8000"
```

**Comment by jasperroloff** [comment]
Hi @Redactlos3,

we had the same problem some months ago and solved it. I just saw your issue here and may help you:

The way, we got OpenSlides running behind Traefik, was to disable (by commenting out) the `front` network & remove the port bindings. Then, we added the (external) `proxy` network to the `proxy` service. This is the network, where the Traefik container belongs to. Though it is defined outside of the scope of this composefile, we still have to tell docker-compose that it exists. Finally, the service needs some Traefik-labels.

Our setup:
- Traefik is running in docker, outside of this project
- There is a `proxy` docker-network, which is connected to the Traefik container
- The `docker-compose.yml` and `.env` files we use, were generated from commit 4f35770769eeb71c22935a885bf82534a2be8934.
- In `.env`, we left `EXTERNAL_HTTP_PORT` empty and only set `INSTANCE_DOMAIN` (and did some email- and feature-related settings)
- Apart from customizing some environment variables unrelated to this issue, we did the following changes to our `docker-compose.yml`:

```
services:
  proxy:
    ...
    networks:
#      - front                    <- comment this line out
      - proxy                     <- add this line ("proxy" is the name of the network which Traefik belongs to)
      - back
#    ports:                       <- comment this line out
#      - "127.0.0.1:8000:8000"    <- comment this line out
    ...
    labels:                       <- add this block
      - "traefik.enable=true"
      - "traefik.http.routers.openslides.entrypoints=http"
      - "traefik.http.routers.openslides.rule=Host(`openslides.example.org`)"
      - "traefik.http.middlewares.openslides-https-redirect.redirectscheme.scheme=https"
      - "traefik.http.routers.openslides.middlewares=openslides-https-redirect"
      - "traefik.http.routers.openslides-secure.entrypoints=https"
      - "traefik.http.routers.openslides-secure.rule=Host(`openslides.example.org`)"
      - "traefik.http.routers.openslides-secure.tls=true"
      - "traefik.http.routers.openslides-secure.tls.certresolver=http"
      - "traefik.http.routers.openslides-secure.service=openslides"
      - "traefik.http.services.openslides.loadbalancer.server.port=8000"
      - "traefik.docker.network=proxy"     <- again, "proxy" is the name of the network which Traefik belongs to
    
  ...

...

networks:
#  front:              <- comment out this line
  proxy:               <- add this line, "proxy" is the name of the network which Traefik belongs to
    external: true     <- add this line
  ...

...
```

As stated above, this config is based on commit 4f35770769eeb71c22935a885bf82534a2be8934. Since then, the `docker-compose.yml`, `.env`, `build.sh` or some docker files may have changed. Also, our Traefik-labels might not be apply to your setup, but this should be easy to solve.

Hope it helps!

Kind regards
Jasper

**Comment by ditschi** [comment]
I got traefik2 working with OpenSlides4. Maybe a similar approach helpes here.

 To get traefik finally working I followed [Configuration of the generated Docker Compose YAML file](https://github.com/OpenSlides/openslides-manage-service#configuration-of-the-generated-docker-compose-yaml-file)
 By configuring  `enableLocalHTTPS: false` in 'my-config.yml'  i can use Traefik without problems.
 
 In the docker-compose.yml I just added the traefik labels section and my external traefik network in the existing `proxy` service (and of course I added  the  external network in the "network" section.)
